store/postgres/src/mapping.rs

```diff
@@ -0,0 +1,1081 @@
+        }
+    }
+
+    fn sql_type(&self) -> &str {
+        match self {
+            ColumnType::Boolean => "boolean",
+            ColumnType::BigDecimal => "numeric",
+            ColumnType::BigInt => "numeric",
+            ColumnType::Bytes => "bytea",
+            ColumnType::Int => "integer",
+            ColumnType::String => "text",
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Column {
+    pub name: SqlName,
+    pub field: String,
+    pub field_type: q::Type,
+    pub column_type: ColumnType,
+    pub derived: Option<String>,
+    pub references: Vec<Reference>,
+}
+
+impl Column {
+    fn new(field: &s::Field, id_type: IdType) -> Result<Column, StoreError> {
+        let sql_name = SqlName::from(&*field.name);
+        Ok(Column {
+            name: sql_name,
+            field: field.name.clone(),
+            column_type: ColumnType::from_value_type(base_type(&field.field_type), id_type)?,
+            field_type: field.field_type.clone(),
+            derived: derived_column(field)?,
+            references: vec![],
+        })
+    }
+
+    pub fn is_reference(&self) -> bool {
+        !self.references.is_empty()
+    }
+
+    pub fn is_primary_key(&self) -> bool {
+        PRIMARY_KEY_COLUMN == self.name.as_str()
+    }
+
+    fn sql_type(&self) -> &str {
+        self.column_type.sql_type()
+    }
+
+    pub fn is_nullable(&self) -> bool {
+        fn is_nullable(field_type: &q::Type) -> bool {
+            match field_type {
+                q::Type::NonNullType(_) => false,
+                _ => true,
+            }
+        }
+        is_nullable(&self.field_type)
+    }
+
+    pub fn is_list(&self) -> bool {
+        fn is_list(field_type: &q::Type) -> bool {
+            use q::Type::*;
+
+            match field_type {
+                ListType(_) => true,
+                NonNullType(inner) => is_list(inner),
+                NamedType(_) => false,
+            }
+        }
+        is_list(&self.field_type)
+    }
+
+    pub fn is_derived(&self) -> bool {
+        self.derived.is_some()
+    }
+
+    fn as_ddl(&self, out: &mut String) -> fmt::Result {
+        write!(out, "    ")?;
+        if self.derived.is_some() {
+            write!(out, " -- ")?;
+        }
+        write!(out, "{:20} {}", self.name, self.sql_type())?;
+        if self.is_list() {
+            write!(out, "[]")?;
+        }
+        if self.name.0 == PRIMARY_KEY_COLUMN || !self.is_nullable() {
+            write!(out, " not null")?;
+        }
+        Ok(())
+    }
+}
+
+/// The name for the primary key column of a table; hardcoded for now
+pub(crate) const PRIMARY_KEY_COLUMN: &str = "id";
+
+pub(crate) const BLOCK_RANGE: &str = "block_range";
+
+#[derive(Clone, Debug)]
+pub struct Table {
+    /// The name of the GraphQL object type ('Thing')
+    pub object: s::Name,
+    /// The name of the database table for this type ('thing'), snakecased
+    /// version of `object`
+    pub name: SqlName,
+
+    pub columns: Vec<Column>,
```

Comment:
There are a bunch of places where the list of columns is filtered for those that are not derived. It seems easy to accidentally treat a derived column as a real column. Might be better to have a separate `derived_columns: Vec<Column>` field.

Reply:
I think there's a whole bunch of things we need to think about wrt how the read interface and SQL query combination will work. Strictly speaking, at this level we don't really need to know about derived columns and references (assuming there is a layer above this that makes sure we treat them correctly)

Maybe we can discuss this tomorrow morning? Would love to get your thoughts on this.

Reply:
Commit efc18827 makes it so that we completely ignore derived fields in the relational schema.

Reply:
Nice, if someone is reading this, we decided to separate the code and data structures between GrapQL -> Read interface and Read interface -> SQL, even if that causes some code duplication at first.